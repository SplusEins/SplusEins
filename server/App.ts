import * as express from 'express';
import * as cookieParser from 'cookie-parser';
import splusController from './controllers/splus';
import mensaController from './controllers/mensa';
import newsController from './controllers/news';
import icsController from './controllers/ics';
import busController from './controllers/bus';

class App {
  public app: express.Application;

  constructor(private basePath: string = '/api') {
    this.app = express();
    this.app.use(cookieParser());
    this.routes();
  }

  private routes(): void {
    this.app.use(this.basePath + '/splus', splusController);
    this.app.use(this.basePath + '/mensa', mensaController);
    this.app.use(this.basePath + '/news', newsController);
    this.app.use(this.basePath + '/ics', icsController);
    this.app.use(this.basePath + '/bus', busController);
  }
}

export default App;
