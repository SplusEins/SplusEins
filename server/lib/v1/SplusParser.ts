import {load} from 'cheerio';
import {ILecture} from './ILecture';

interface IBlock {
    title: string;
    durationHours: number;
    durationSlots: number;
    info: string;
    room: string;
    lecturer: string
}

export class SplusParser {
    private readonly startHour = 7;

    private readonly $: CheerioStatic;
    private readonly $table: Cheerio;
    private readonly $rows: Cheerio;

    private colWidths: number[] = [];
    private colBlocked: number[] = [];

    private blocks: IBlock[][][] = [];

    constructor(data: string) {
        this.$ = load(data);

        this.$table = this.$(this.$('table.grid-border-args')[0]);
        this.$rows = this.$table.find('> tbody > tr');

        this.parseColWidths();
        this.parseTable();
    }

    getLectures(): ILecture[] {
        let lectures: ILecture[] = [];

        for (let i = 0; i < this.blocks.length; i++) {
            const begin = this.startHour + i * 0.25;

            const slot = this.blocks[i];
            for (let j = 0; j < slot.length; j++) {
                const day = slot[j];
                for (let k = 0; k < day.length; k++) {
                    const block = day[k];

                    const lecture = {
                        title: block.title,
                        day: j,
                        begin: begin,
                        end: begin + block.durationHours,
                        info: block.info,
                        room: block.room,
                        lecturer: block.lecturer
                    };

                    lectures.push(lecture);
                }
            }
        }

        return lectures;
    }

    private getDay(col: number) {
        return this.colWidths.findIndex(width => (col -= width) < 0);
    }

    private parseColWidths() {
        const $cols = this.$(this.$rows[0]).find('td:not(:first-child)');

        for (let i = 0; i < $cols.length; i++) {
            const $col = this.$($cols[i]);
            this.colWidths[i] = parseInt($col.attr('colspan'));
        }

        let index = 0;
        for (let i = 0; i < this.colWidths.length; i++) {
            for (let j = 0; j < this.colWidths[i]; i++) {
                this.colBlocked[index++] = 0;
            }
        }
    }

    private parseTable() {
        for (let i = 0; i < this.$rows.length - 1; i++) {
            // i + 1 to skip the columns row
            const $row = this.$(this.$rows[i + 1]);

            this.colBlocked = this.colBlocked.map(v => v > 0 ? v - 1 : 0);

            this.blocks.push(this.parseRow($row));
        }
    }

    private parseRow($row: Cheerio) {
        const $cols = $row.find('> td:not(:first-child)');
        const blocks: IBlock[][] = [];

        let offset = 0;
        for (let i = 0; i < $cols.length; i++) {
            while (this.colBlocked[i + offset] > 0) {
                const day = this.getDay(i + offset);
                if (blocks[day] === undefined) {
                    blocks[day] = [];
                }

                offset++;
            }

            const day = this.getDay(i + offset);
            if (blocks[day] === undefined) {
                blocks[day] = [];
            }

            const $c = this.$($cols[i]);
            if ($c.hasClass('object-cell-border')) {
                const block = this.parseBlock($c);
                blocks[day].push(block);

                this.colBlocked[i + offset] = block.durationSlots;
            }
        }

        return blocks;
    }

    private parseBlock($block: Cheerio): IBlock {
        const $tblTitle = $block.find('table:nth-child(1)');
        const $tblInfo = $block.find('table:nth-child(2)');
        const $tblLocation = $block.find('table:nth-child(3)');

        const $tdRoom = $tblLocation.find('td:nth-child(1)');
        const $tdLecturer = $tblLocation.find('td:nth-child(2)');

        // Remove the Modulhandbuch "mhb" link
        $tblTitle.find('br + strong').remove();

        const slots = parseInt($block.attr('rowspan'));
        return {
            title: $tblTitle.text().trim(),
            durationHours: slots / 4,
            durationSlots: slots,
            info: $tblInfo.text().trim(),
            room: $tdRoom.text().trim(),
            lecturer: $tdLecturer.text().trim()
        };
    }
}
