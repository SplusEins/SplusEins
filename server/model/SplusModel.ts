export interface ParsedBlock {
    title: string;
    durationHours: number;
    durationSlots: number;
    info: string;
    room: string;
    lecturer: string
}

export interface ParsedLecture {
    title: string;
    start: Date;
    end: Date;
    duration: number;
    info: string;
    room: string,
    lecturer: string;
}
