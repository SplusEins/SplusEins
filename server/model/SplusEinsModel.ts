import { ParsedLecture } from './SplusModel';
import * as crypto from 'crypto';

export interface NewsElement {
  title: string;
  link: string;
  text: string;
  date?: Date;
  source?: string;
};

export interface MensaDayPlan {
  date: Date;
  data: object;
};

export interface TimetableRequest {
  id: string;
  week: number;
  setplan: boolean;
  raumplan: boolean;
};

export interface EventMetadata {
  organiserShortname: string;
  organiserName: string;
  description: string;
};

export interface TimetableMetadata {
  splusID: string | string[];
  faculty: string | string[];
  degree: string | string[];
  specialisation: string | string[];
  semester: number | number [];
}

export interface Timetable {
  name: string,
  events: Event[];
  meta: TimetableMetadata;
}

export class Event {
  id: string;
  title: string;
  start: Date;
  end: Date;
  duration: number;
  location: string;
  meta: EventMetadata;

  private generateId(title: string): string {
    //since the id is part of the url it should not contain /, + or =
    return crypto.createHash('sha1').update(title).digest('base64').replace(/[\/+=]/g, '').slice(0, 5);
  }

  private generateOrganiserShortname(lecturer: string): string {
    return lecturer
      .replace(/Prof|Dr|Dipl|Ing|Herr|Frau|MA|BA/g, '')
      .replace('ä', 'ae').replace('ö', 'oe').replace('ü', 'ue').replace('ß', 'ss')
      .replace(/[^a-z]/gi, ' ')
      .split(' ')
      .filter((w) => w.length > 3)
      .join(' ');
  }

  constructor(lecture: ParsedLecture) {
    this.id = this.generateId(lecture.title);
    this.title = lecture.title;
    this.start = lecture.start;
    this.end = lecture.end;
    this.duration = lecture.duration;
    this.location = lecture.room;
    this.meta = {
                 organiserShortname: this.generateOrganiserShortname(lecture.lecturer),
                 organiserName: lecture.lecturer,
                 description: lecture.info
                };
  }
};
