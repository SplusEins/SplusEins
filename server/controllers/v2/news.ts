import * as express from 'express';
import * as cors from 'cors';
import * as cacheManager from 'cache-manager';
import * as fsStore from 'cache-manager-fs-hash';
import * as moment from 'moment';
import * as cheerio from 'cheerio';
import fetch from 'node-fetch';
import { URLSearchParams } from 'url';

import { NewsElement } from '../../model/v2/SplusEinsModel';

// default must be in /tmp because the rest is RO on AWS Lambda
const CACHE_PATH = process.env.CACHE_PATH || '/tmp/spluseins-cache';
const CACHE_DISABLE = !!process.env.CACHE_DISABLE;
const CACHE_SECONDS = parseInt(process.env.NEWS_CACHE_SECONDS || '1800');

const router = express.Router();
const cache = CACHE_DISABLE ?
  cacheManager.caching({ store: 'memory', max: 0 }) :
  cacheManager.caching({
    store: fsStore,
    options: {
      path: CACHE_PATH,
      ttl: 60,
      subdirs: true,
    },
  });

/**
 * Accept CORS preflight requests.
 */
router.options('/ostfalia', cors());
router.options('/ostfalia/:faculty', cors());
router.options('/campus38', cors());

/**
 * Get Ostfalia news
 * @returns NewsElement[]
 */
router.get('/ostfalia', cors(), async (req, res, next) => {
  const key = 'ostfalia-news';

  try {
    const data = await cache.wrap(key, async () => {
      console.log(`ostfalia news cache miss for key ${key}`);

      const query = new URLSearchParams();
      query.append('itemsPerPage', '10');
      query.append('collectorParam',
        'fq=type:of-news' +
        '&fq=parent-folders:"/sites/default/de/campus/.content/newsentries10/"' +
        '&sort=newsdate_de_dt desc' +
        '|createPath=/sites/default/de/campus/.content/newsentries10/news_%(number).xml');
      query.append('showDate', 'true');
      query.append('currPage', '1');

      const response = await fetch(
        'https://www.ostfalia.de/cms/system/modules/de.ostfalia.module.template/elements/renderNewsList.jsp', {
          method: 'POST',
          body: query,
        }).then((res) => res.text());

      const $ = cheerio.load(response);
      return $('article').map(function(i, article) {
        return <NewsElement> {
          title: $('a', this).text().trim(),
          link: 'https://www.ostfalia.de' + $('a', this).attr('href'),
          text: $('p', this).last().text().trim(),
          date: moment($('p', this).first().text().trim(), 'DD.MM.YY').utcOffset('+0100').toDate(),
        };
      }).get();
    }, { ttl: CACHE_SECONDS });

    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.json(data);
  } catch (error) {
    next(error);
  }
});

/**
 * Get Ostfalia faculty news
 * @returns NewsElement[]
 */
router.get('/ostfalia/:faculty', cors(), async (req, res, next) => {
  const faculty = req.params.faculty;
  if (!['i', 'r', 'e', 'sz', 'wf', 'wob', 'sud'].includes(faculty)) {
    // Informatik, Recht, Elektrotechnik,
    // Salzgitter, WolfenbÃ¼ttel, Wolfsburg, Suderburg
    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.sendStatus(404);
    return;
  }
  const key = 'ostfalia-news-' + faculty;

  try {
    let url;
    if (faculty.length == 1) {
      url = 'https://www.ostfalia.de/cms/de/' + faculty;
    } else {
      url = 'https://www.ostfalia.de/cms/de/campus/' + faculty;
    }

    const data = await cache.wrap(key, async () => {
      console.log(`ostfalia faculty news cache miss for key ${key}`);

      const response = await fetch(url)
        .then((res) => res.text());
      const $ = cheerio.load(response);

      if (faculty == 'e') {
        return $('article').eq(1).find('.ostfalia-content div').map(function(i, paragraph) {
          const p = $('p', this).last();
          const text = p.text();
          return <NewsElement> {
            title: text.substring(text.indexOf(':') + 1,
              text.indexOf('mehr')).trim(),
            link: 'https://www.ostfalia.de' + $('a', p).attr('href'),
            text: '',
          }
        }).get();
      }

      if (faculty == 'i') {
        return $('article .ostfalia-content table tbody td').map(function(i, td) {
          let link = $('a', this).first().attr('href');
          if (!!link && !link.startsWith('http')) {
            link = 'https://www.ostfalia.de' + link;
          } else {
            link = '';
          }

          return <NewsElement> {
            title: $(this).text().trim(),
            link,
            text: '',
          }
        }).get();
      }

      return $('article.news-campus').map(function(i, article) {
        return <NewsElement> {
          title: $('a', this).text().trim(),
          link: 'https://www.ostfalia.de' + $('a', this).attr('href'),
          text: $('p', this).last().text().trim(),
          date: moment($('p', this).first().text().trim(), 'DD.MM.YY').utcOffset('+0100').toDate(),
        };
      }).get();
    }, { ttl: CACHE_SECONDS });

    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.json(data);
  } catch (error) {
    next(error);
  }
});

/**
 * Get Campus38 news
 * @returns NewsElement[]
 */
router.get('/campus38', cors(), async (req, res, next) => {
  const key = 'campus38-news';

  try {
    const data = await cache.wrap(key, async () => {
      console.log(`campus38 news cache miss for key ${key}`);

      const response = await fetch('https://www.campus38.de/newsfeed.xml')
        .then((res) => res.text());
      const $ = cheerio.load(response, { xmlMode: true });
      return $('entry').map(function(i, article) {
        return <NewsElement> {
          title: $('title', this).text().trim(),
          link: $('link', this).attr('href'),
          text: $('summary', this).text().trim(),
          date: moment($('published', this).text().trim().split('T')[0]).utcOffset('+0100').toDate(),
        };
      }).slice(0, 10).get();
    }, { ttl: CACHE_SECONDS });

    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.json(data);
  } catch (error) {
    next(error);
  }
});

export default router;
