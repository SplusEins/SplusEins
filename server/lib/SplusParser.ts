import { load } from 'cheerio';
import { ParsedLecture, ParsedBlock } from '../model/SplusModel';
import * as moment from 'moment-timezone'

export class SplusParser {
    private readonly startHour = 7;

    private readonly $: CheerioStatic;
    private readonly $table: Cheerio;
    private readonly $rows: Cheerio;

    private colWidths: number[] = [];
    private colBlocked: number[] = [];

    private blocks: ParsedBlock[][][] = [];

    constructor(data: string) {
        this.$ = load(data);

        this.$table = this.$(this.$('table.grid-border-args')[0]);
        this.$rows = this.$table.find('> tbody > tr');

        this.parseColWidths();
        this.parseTable();
    }

    getLectures(weekOfYear: number): ParsedLecture[] {
        let lectures: ParsedLecture[] = [];

        for (let i = 0; i < this.blocks.length; i++) {
            const begin = this.startHour + i * 0.25;

            const slot = this.blocks[i];
            for (let j = 0; j < slot.length; j++) {
                const day = slot[j];
                for (let k = 0; k < day.length; k++) {
                    const block = day[k];

                    const date = this.getDate(j, begin, block.durationHours, weekOfYear);
                    const lecture : ParsedLecture= {
                        title: block.title,
                        start: date.start,
                        end: date.end,
                        duration: block.durationHours,
                        info: block.info,
                        room: block.room,
                        lecturer: block.lecturer
                    };

                    lectures.push(lecture);
                }
            }
        }

        return lectures;
    }

    private getDate(lectureDay, start: number, duration: number, week: number,): {start: Date, end: Date} {
        const day = moment()
            .tz('Europe/Berlin')
            .startOf('date')
            .isoWeek(week % 52)
            .isoWeekday(lectureDay + 1)

        return {start: day.add(start, 'hours').toDate(), end: day.add(duration, 'hours').toDate()}
    }

    private getDay(col: number) {
        return this.colWidths.findIndex(width => (col -= width) < 0);
    }

    private parseColWidths() {
        const $cols = this.$(this.$rows[0]).find('td:not(:first-child)');

        for (let i = 0; i < $cols.length; i++) {
            const $col = this.$($cols[i]);
            this.colWidths[i] = parseInt($col.attr('colspan'));
        }

        let index = 0;
        for (let i = 0; i < this.colWidths.length; i++) {
            for (let j = 0; j < this.colWidths[i]; i++) {
                this.colBlocked[index++] = 0;
            }
        }
    }

    private parseTable() {
        for (let i = 0; i < this.$rows.length - 1; i++) {
            // i + 1 to skip the columns row
            const $row = this.$(this.$rows[i + 1]);

            this.colBlocked = this.colBlocked.map(v => v > 0 ? v - 1 : 0);

            this.blocks.push(this.parseRow($row));
        }
    }

    private parseRow($row: Cheerio) {
        const $cols = $row.find('> td:not(:first-child)');
        const blocks: ParsedBlock[][] = [];

        let offset = 0;
        for (let i = 0; i < $cols.length; i++) {
            while (this.colBlocked[i + offset] > 0) {
                const day = this.getDay(i + offset);
                if (blocks[day] === undefined) {
                    blocks[day] = [];
                }

                offset++;
            }

            const day = this.getDay(i + offset);
            if (blocks[day] === undefined) {
                blocks[day] = [];
            }

            const $c = this.$($cols[i]);
            if ($c.hasClass('object-cell-border')) {
                const block = this.parseBlock($c);
                blocks[day].push(block);

                this.colBlocked[i + offset] = block.durationSlots;
            }
        }

        return blocks;
    }

    private parseBlock($block: Cheerio): ParsedBlock {
        const $tblTitle = $block.find('table:nth-child(1)');
        const $tblInfo = $block.find('table:nth-child(2)');
        const $tblLocation = $block.find('table:nth-child(3)');

        const $tdRoom = $tblLocation.find('td:nth-child(1)');
        const $tdLecturer = $tblLocation.find('td:nth-child(2)');

        // Remove the Modulhandbuch "mhb" link
        $tblTitle.find('br + strong').remove();

        const slots = parseInt($block.attr('rowspan'));
        return {
            title: $tblTitle.text().trim(),
            durationHours: slots / 4,
            durationSlots: slots,
            info: $tblInfo.text().trim(),
            room: $tdRoom.text().trim(),
            lecturer: $tdLecturer.text().trim()
        };
    }
}
