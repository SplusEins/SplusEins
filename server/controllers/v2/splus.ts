import * as express from 'express';
import * as cors from 'cors';
import * as TIMETABLES from '../../../assets/timetables.json';

import { TimetableRequest, TimetableMetadata, Timetable } from '../../model/v2/SplusEinsModel';
import getEvents from '../../lib/v2/SplusApi';

const CACHE_SECONDS = parseInt(process.env.SPLUS_CACHE_SECONDS || '10800');

const router = express.Router();

/**
 * Accept CORS preflight requests.
 */
router.options('/:timetable/:week', cors());
router.options('/:name/:timetables/:week/:lectures?', cors());

/**
 * Get Timetable for given splusId and week
 *
 * @param timetable The splus "identifier" query param without "#" prefix.
 * @param week The splus "week" request param. ISO week of year below 52 or week of next year above 52.
 * @return Timetable
 */
router.get('/:timetable/:week', cors(), async (req, res, next) => {

  const timetableId = req.params.timetable;
  const requestedTimetable = TIMETABLES.find(({ id }) => id == timetableId);

  if (!requestedTimetable) {
    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.sendStatus(404);
    return;
  }

  const week = parseInt(req.params.week);

  try {
    const request: TimetableRequest = <TimetableRequest> {id: requestedTimetable.id, week: week, setplan: requestedTimetable.setplan};
    const events = await getEvents([ request ]);

    const meta : TimetableMetadata = <TimetableMetadata> {
      splusID: requestedTimetable.id,
      faculty: requestedTimetable.faculty,
      degree: requestedTimetable.degree,
      specialisation: requestedTimetable.label,
      semester: Number(requestedTimetable.semester)
    }
    const timetable: Timetable = <Timetable> {
      name: `${(requestedTimetable.degree)} ${requestedTimetable.label} - ${requestedTimetable.semester}. Semester`,
      events: events,
      meta: meta
    }

    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.json(timetable);
  } catch (error) {
    next(error);
  }
});


/**
 * Get Timetable with given name for given splusIds, week und lecture Ids
 *
 * @param timetables Comma-separated list of timetable IDs
 * @param week The splus "week" request param. ISO week of year below 52 or week of next year above 52.
 * @param lectures Comma-separated list of lecture title IDs
 * @param name Name of requested Timetable
 * @return Timetable
 */
router.get('/:timetables/:week/:lectures?/:name', async (req, res, next) => {

  const timetableIds = <string[]>req.params.timetables.split(',');
  const titleIds = <string[]>(req.params.lectures || '')
    .split(',')
    .filter((titleId) => titleId.length > 0);

  const timetables = timetableIds
    .map((timetableId) => TIMETABLES.find(({ id }) => id == timetableId))
    .filter((timetable) => timetable != undefined);

  if (timetables.length == 0) {
    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.sendStatus(404);
    return;
  }

  const name = req.params.name;
  const week = parseInt(req.params.week);

  try {
    const requests: TimetableRequest[] = [];
    timetables.forEach((timetable) => requests.push( <TimetableRequest> {id: timetable.id, week: week, setplan: timetable.setplan}));

    const allEvents = await getEvents(requests);
    const filteredEvents = titleIds.length > 0 ?
      allEvents.filter(({ id }) => titleIds.includes(id))
      : allEvents;

    const meta: TimetableMetadata = <TimetableMetadata> {
      splusID: timetableIds,
      faculty: timetables.map((x) => x.faculty),
      degree: timetables.map((x) => x.degree),
      specialisation: timetables.map((x) => x.label),
      semester: timetables.map((x) => Number(x.semester))
    };

    const timetable: Timetable = <Timetable> {
      name: name,
      events: filteredEvents,
      meta: meta
    }

    res.set('Cache-Control', `public, max-age=${CACHE_SECONDS}`);
    res.json(timetable);
  } catch (error) {
    next(error);
  }
});

export default router;
