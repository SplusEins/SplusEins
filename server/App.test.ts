process.env.CACHE_DISABLE = '1';

import App from './App';
import * as request from 'supertest';

import { SplusParser } from './lib/v2/SplusParser';
import { readFile } from 'fs';
import { promisify } from 'util';
import { Event } from './model/v2/SplusEinsModel';
import { ParsedLecture } from './model/v2/SplusModel';

async function splusApiMock(identifier: string, weekOfYear: string) {
  const htmlPath = './server/__snapshots__/splus_ibi1_44.html';
  const html = await promisify(readFile)(htmlPath, 'utf8');
  const lectures: ParsedLecture[] = new SplusParser(html).getLectures(12);
  return lectures.map((lecture : ParsedLecture) => new Event(lecture));
}

jest.mock('./lib/v2/SplusApi', () => ({
  default: jest.fn().mockImplementation(splusApiMock),
  getEvents: jest.fn().mockImplementation(splusApiMock),
}));

describe('Test backend', () => {
  const app = new App().app;

  it('should return parsed lectures', async () => {
    const response = await request(app).get(
      '/api/v2/splus/SPLUS7A3292/10');
    expect(response.body).toMatchSnapshot();
  });

  it('should create a ICS for multiple timetables and filters', async () => {
    const response = await request(app).get(
      '/api/ics/v1/SPLUS7A3292,SPLUS7A3293/GdPL,DSB');
    expect(response.statusCode).toBe(200);
    expect(response.header['content-type']).toBe('text/plain; charset=utf-8');
    expect(response.text).toContain('BEGIN:VCALENDAR');
    expect(response.text).toContain('BEGIN:VEVENT');
    expect(response.text).toContain('Diskrete Strukturen');
    expect(response.text).toContain('Grundlagen des Programmierens - Labor');
    expect(response.text).not.toContain('Technische Grundlagen der Informatik');
  });

  it('should create an ICS for one timetable and one filter', async () => {
    const response = await request(app).get(
      '/api/ics/v1/SPLUS7A3292/GdPL');
    expect(response.statusCode).toBe(200);
    expect(response.header['content-type']).toBe('text/plain; charset=utf-8');
    expect(response.text).toContain('BEGIN:VCALENDAR');
    expect(response.text).toContain('BEGIN:VEVENT');
    expect(response.text).toContain('Grundlagen des Programmierens - Labor');
    expect(response.text).not.toContain('Technische Grundlagen der Informatik');
  });

  it('should create an ICS for one timetable and no filter', async () => {
    const response = await request(app).get(
      '/api/ics/v1/SPLUS7A3292');
    expect(response.statusCode).toBe(200);
    expect(response.header['content-type']).toBe('text/plain; charset=utf-8');
    expect(response.text).toContain('BEGIN:VCALENDAR');
    expect(response.text).toContain('BEGIN:VEVENT');
    expect(response.text).toContain('Grundlagen des Programmierens - Labor');
    expect(response.text).toContain('Technische Grundlagen der Informatik');
  });

  it('should not find an ICS for a timetable that does not exist', async () => {
    const response = await request(app).get(
      '/api/ics/v1/DOESNOTEXIST');
    expect(response.statusCode).toBe(404);
  });

  it('should return an ICS matching the snapshot', async () => {
    const response = await request(app).get(
      '/api/ics/v1/SPLUS7A3292,SPLUS7A3293/GdPL,DSB');
    const deterministicText = response.text.replace(/^DTSTAMP:.*$/gm, '');
    expect(deterministicText).toMatchSnapshot();
  });
});
