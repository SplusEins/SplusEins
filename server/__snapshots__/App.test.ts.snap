// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test backend should return parsed lectures 1`] = `
Array [
  Object {
    "begin": 7,
    "day": 2,
    "duration": 13.25,
    "id": "Reformationstag - ",
    "info": "",
    "lecturer": "",
    "room": "",
    "start": "2018-10-03T00:00:00.000Z",
    "title": "Reformationstag",
  },
  Object {
    "begin": 8.25,
    "day": 1,
    "duration": 1.5,
    "id": "Grundlagen des Programmierens - Labor - Prof. Dr. F. Höppner, M.Sc. F. Morgenstern",
    "info": "",
    "lecturer": "Prof. Dr. F. Höppner, M.Sc. F. Morgenstern",
    "room": "P1, P4, P3, P2",
    "start": "2018-10-02T00:00:00.000Z",
    "title": "Grundlagen des Programmierens - Labor",
  },
  Object {
    "begin": 8.25,
    "day": 3,
    "duration": 1.5,
    "id": "Mathe Repetitorium - Sondertermin - BInf - ",
    "info": "Dr. S. Bellmer, Raum 252. Dieser Sondertermin findet zusammen mit Mathe-Repetitorium für Wirtschaftsinformatik statt!",
    "lecturer": "",
    "room": "",
    "start": "2018-10-04T00:00:00.000Z",
    "title": "Mathe Repetitorium - Sondertermin - BInf",
  },
  Object {
    "begin": 8.25,
    "day": 4,
    "duration": 1.5,
    "id": "Grundlagen des Programmierens - Gruppe B - Prof. Dr. F. Höppner",
    "info": "",
    "lecturer": "Prof. Dr. F. Höppner",
    "room": "Hörsaal 223",
    "start": "2018-10-05T00:00:00.000Z",
    "title": "Grundlagen des Programmierens - Gruppe B",
  },
  Object {
    "begin": 8.25,
    "day": 4,
    "duration": 1.5,
    "id": "Grundlagen des Programmierens - Gruppe A - Prof. Dr. M. Huhn",
    "info": "",
    "lecturer": "Prof. Dr. M. Huhn",
    "room": "Hörsaal 026",
    "start": "2018-10-05T00:00:00.000Z",
    "title": "Grundlagen des Programmierens - Gruppe A",
  },
  Object {
    "begin": 9.5,
    "day": 0,
    "duration": 3.25,
    "id": "I-Einführung und Kompetenzen der Informatik (CS) - Gruppe 1 - M.A. J. Schwiebert (CS)",
    "info": "",
    "lecturer": "M.A. J. Schwiebert (CS)",
    "room": "Seminarraum 152",
    "start": "2018-10-01T00:00:00.000Z",
    "title": "I-Einführung und Kompetenzen der Informatik (CS) - Gruppe 1",
  },
  Object {
    "begin": 9.5,
    "day": 0,
    "duration": 3.25,
    "id": "Einführung und Kompetenzen der Informatik (CS) - Prof. Dr.-Ing. N. Jensen",
    "info": "",
    "lecturer": "Prof. Dr.-Ing. N. Jensen",
    "room": "11/1",
    "start": "2018-10-01T00:00:00.000Z",
    "title": "Einführung und Kompetenzen der Informatik (CS)",
  },
  Object {
    "begin": 10,
    "day": 1,
    "duration": 1.5,
    "id": "Grundlagen des Programmierens - Labor - Prof. Dr. F. Höppner, M.Sc. F. Morgenstern",
    "info": "",
    "lecturer": "Prof. Dr. F. Höppner, M.Sc. F. Morgenstern",
    "room": "P1, P4, P3, P2",
    "start": "2018-10-02T00:00:00.000Z",
    "title": "Grundlagen des Programmierens - Labor",
  },
  Object {
    "begin": 10,
    "day": 3,
    "duration": 1.5,
    "id": "Mathe Cafe 1. Sem - Dr. S. Bellmer",
    "info": "",
    "lecturer": "Dr. S. Bellmer",
    "room": "Hörsaal 252",
    "start": "2018-10-04T00:00:00.000Z",
    "title": "Mathe Cafe 1. Sem",
  },
  Object {
    "begin": 10,
    "day": 4,
    "duration": 1.5,
    "id": "Diskrete Strukturen-Bainf - Prof. Dr. P. Riegler",
    "info": "",
    "lecturer": "Prof. Dr. P. Riegler",
    "room": "Scale-Up-Room 201, I-WF-Ex-7/? (Alter Speisesaal Cafe Limes)",
    "start": "2018-10-05T00:00:00.000Z",
    "title": "Diskrete Strukturen-Bainf",
  },
  Object {
    "begin": 12,
    "day": 1,
    "duration": 1.5,
    "id": "Diskrete Strukturen-Bainf - Prof. Dr. P. Riegler",
    "info": "Die VL findet in 7/? statt.",
    "lecturer": "Prof. Dr. P. Riegler",
    "room": "Scale-Up-Room 201, I-WF-Ex-7/? (Alter Speisesaal Cafe Limes)",
    "start": "2018-10-02T00:00:00.000Z",
    "title": "Diskrete Strukturen-Bainf",
  },
  Object {
    "begin": 12,
    "day": 4,
    "duration": 1.5,
    "id": "Informatik-Lounge 1.Sem - Dipl.-Inf. D. Dick-Anwander",
    "info": "",
    "lecturer": "Dipl.-Inf. D. Dick-Anwander",
    "room": "Poolraum 132 (P8)",
    "start": "2018-10-05T00:00:00.000Z",
    "title": "Informatik-Lounge 1.Sem",
  },
  Object {
    "begin": 14,
    "day": 3,
    "duration": 4,
    "id": "Rhetorik und wissenschaftliches Arbeiten - Gruppe 2 (CS) - ",
    "info": "",
    "lecturer": "",
    "room": "Seminarraum 152",
    "start": "2018-10-04T00:00:00.000Z",
    "title": "Rhetorik und wissenschaftliches Arbeiten - Gruppe 2 (CS)",
  },
  Object {
    "begin": 14,
    "day": 3,
    "duration": 4,
    "id": "Rhetorik und wissenschaftliches Arbeiten - Gruppe 3 (CS) - ",
    "info": "Dozenten: M.A. J. Schwiebert / Dipl.-Ing. H. Perkuhn",
    "lecturer": "",
    "room": "Seminarraum 136",
    "start": "2018-10-04T00:00:00.000Z",
    "title": "Rhetorik und wissenschaftliches Arbeiten - Gruppe 3 (CS)",
  },
  Object {
    "begin": 14.25,
    "day": 0,
    "duration": 1.5,
    "id": "Technische Grundlagen der Informatik Alt: Technik der Informatik-VL - Dipl.-Ing. K. Dammann",
    "info": "",
    "lecturer": "Dipl.-Ing. K. Dammann",
    "room": "11/1",
    "start": "2018-10-01T00:00:00.000Z",
    "title": "Technische Grundlagen der Informatik Alt: Technik der Informatik-VL",
  },
  Object {
    "begin": 14.25,
    "day": 1,
    "duration": 1.5,
    "id": "Technische Grundlagen der Informatik Alt: Technik der Informatik-VL - Dipl.-Ing. K. Dammann",
    "info": "",
    "lecturer": "Dipl.-Ing. K. Dammann",
    "room": "Hörsaal 252",
    "start": "2018-10-02T00:00:00.000Z",
    "title": "Technische Grundlagen der Informatik Alt: Technik der Informatik-VL",
  },
]
`;
